Add demo script (#3)
oktadeveloper/okta-spring-boot-docker-buildpacks-example: demo.adoc

----
curl https://start.spring.io/starter.tgz -d dependencies=web,okta \
   -d bootVersion=2.5.4 \
   -d groupId=com.okta \
   -d artifactId=demospringboot \
   -d type=gradle-project \
   -d language=kotlin \
   -d baseDir=springboot-docker-demo | tar -xzvf -
----

. Update the `DemoApplication` class to allow authenticated access.
+
[source,kotlin]
----
@Configuration
class OktaOAuth2WebSecurityConfigurerAdapter: WebSecurityConfigurerAdapter() {
    override fun configure(http: HttpSecurity) {
        http.authorizeRequests().anyRequest().permitAll()
    }
}
----

. Add a controller that welcomes the user.
+
[source,kotlin]
----
@RestController
class WebController {
    @RequestMapping("/")
    fun home(user: Principal?) = "Welcome, ${user?.name ?: "guest"}!"
}
----

=== Run Your Spring Boot Application

. Open a terminal and start your app.
+
[source,shell]
----
./gradlew bootRun
----

. Open your favorite browser and navigate to `http://localhost:8080`. You will see a greeting for the guest user.

== Build a Spring Boot Docker Image

. Build your app and send the image to your local Docker daemon:
+
[source,shell]
----
./gradlew bootBuildImage --imageName=springbootdemo
----
+

> NOTE: The `--imageName` parameter allows specifying an image name. Without it, the name would be something like `demospringboot:0.0.1-SNAPSHOT`.

Reviewer: ```suggestion
NOTE: The `--imageName` parameter allows you to specify an image name. Without it, the name would be something like `demospringboot:0.0.1-SNAPSHOT`.
```
clarifying

. Start your containerized web application with Docker:
+
[source,shell]
----
docker run -it -p8080:8080 springbootdemo
----

. Make sure it works at `http://localhost:8080`.

== Secure Your Spring Boot Application in Docker

User management is never an easy task and, most certainly, is not the main objective of your application. Okta is an identity provider that helps you to take care of routine work such as:

- Implementing OAuth 2.0

> - Adding Social Login

Reviewer: ```suggestion
- Adding social login
```
lowercase since it's not a product name or proper noun...

- Logout: `http://localhost:8080`
. You'll need to run `source .okta.env` to set these values as environment variables.

TIP: If you're on Windows, rename the file to `.okta.bat` and change `export` to `set`.

=== Configure Spring Security to Lock Down Access

. Modify your security settings in `DemoApplication` to allow access for authorized users only.
+
[source,kotlin]
----
@Configuration
class OktaOAuth2WebSecurityConfigurerAdapter: WebSecurityConfigurerAdapter() {
    override fun configure(http: HttpSecurity) {
        http.authorizeRequests().anyRequest().authenticated();
    }
}
----
+
That's it. The Okta Spring Boot Starter takes care of the rest!

. Rebuild the application:
+
[source,shell]
----
./gradlew bootBuildImage --imageName=springbootdemo
----

=== Start Spring Boot Application in Docker

. Start your application in Docker with your Okta settings as environment variables:
+
[source,shell]
----
docker run -it -p8080:8080 \
  -e OKTA_OAUTH2_ISSUER="https://dev-xxxxxx.okta.com/oauth2/default" \
  -e OKTA_OAUTH2_CLIENT_SECRET="yyyyyyyyyyyyyyyyyyy" \
  -e OKTA_OAUTH2_CLIENT_ID="zzzzzzzzzzzzzzzz" \
  springbootdemo
----
+

> The argument `-e` allows to set an environment variable for the application running _inside_ your container and `-p` maps container's ports to `localhost` ports.

Reviewer: ```suggestion
The argument `-e` allows you to set an environment variable for the application running _inside_ your container and `-p` maps the container's ports to `localhost` ports.
```
adding object (you), adding article (the)

. Navigate to `http://localhost:8080` in your browser and log in using your Okta credentials.

. If you want to print the user's name (instead of the `sub` claim), use `@AuthenticatedPrincipal` and `OidcUser`:
+
[source,kotlin]
----
fun home(@AuthenticationPrincipal user: OidcUser?) = "Welcome, ${user?.fullName ?: "guest"}!"
----

> == Use a dotenv File for Environment Variables

Reviewer: question: do you ever use code formatting in headers? `dotenv` ??

Author: Yes, I do. dotenv is a library name though, and I rarely put those in backticks. https://www.npmjs.com/package/dotenv

> Using environment variables in the command-line works, but it's not very convenient and can leave unwanted traces of the secrets in your terminal history. Docker supports dotenv file format, which makes it easier to set multiple environment parameters.

Reviewer: ```suggestion
Using environment variables in the command line works, but it's not very convenient and can leave unwanted traces of the secrets in your terminal history. Docker supports the dotenv file format, which makes it easier to set multiple environment parameters.
```
no hyphen needed for command line, since it's not modifying another word.

+
[source,properties]
----
server.port=${PORT:8080}
----

. Then, build your image with `--builder heroku/spring-boot-buildpacks`:
+
[source,shell]
----
./gradlew bootBuildImage --imageName=springbootdemo \
  --builder heroku/spring-boot-buildpacks
----

. Install the https://devcenter.heroku.com/articles/heroku-cli[Heroku CLI] and create an app on Heroku:

    heroku create

. Log in to Heroku's container registry and push your app:
+
[source,shell]
----
heroku container:login
docker tag springbootdemo registry.heroku.com/<your-app-name>/web
docker push registry.heroku.com/<your-app-name>/web
----

. Set your Okta OIDC app settings as environment variables:
+
[source,shell]
----
heroku config:set \
  OKTA_OAUTH2_ISSUER="https://{yourOktaDomain}/oauth2/default" \
  OKTA_OAUTH2_CLIENT_ID="{clientId}" \
  OKTA_OAUTH2_CLIENT_SECRET="{clientSecret}"
----

. Next, release your container and tail the logs:
+
[source,shell]
----
heroku container:release web
heroku logs --tail
----

. Run `heroku open` to open your app and sign in. Witness a 400 error for an invalid redirect URI.

> . Update your Okta OIDC app to have your Heroku app's redirect URIs:

Reviewer: ```suggestion
. Update your Okta OIDC app to use your Heroku app's redirect URIs:
```
more precise
